src/main/java/com/bookstore/service/order/OrderServiceImpl.java

```diff
@@ -0,0 +1,117 @@
+    private final AuthentificationService authentificationService;
+
+    @Transactional
+    @Override
+    public List<OrderResponseDto> getAll(Pageable pageable) {
+        return orderRepository.findAllByUserId(authentificationService.getUserId()).stream()
+                .map(orderMapper::toDto)
+                .toList();
+    }
+
+    @Transactional
+    @Override
+    public OrderResponseDto save(CreateOrderRequestDto requestDto) {
+        User user = (User) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
+        ShoppingCart cart = shoppingCartService.getById(user.getId());
+        Order order = new Order();
+        order.setUser(user);
+        order.setShippingAddress(requestDto.getShippingAddress());
+        order.setOrderDate(LocalDateTime.now());
+        order.setTotal(cart.getCartItems().stream()
+                .map(c -> c.getBook().getPrice().multiply(BigDecimal.valueOf(c.getQuantity())))
+                .reduce(BigDecimal.valueOf(0), BigDecimal::add));
+        orderRepository.save(order);
+        OrderResponseDto dto = orderMapper.toDto(order);
+        dto.setOrderItems(getOrderItemDtos(getOrderItems(cart, order)));
+        return dto;
+    }
+
+    @Transactional
+    @Override
+    public OrderResponseDto updateOrderStatus(Long id, UpdateStatusRequestDto requestDto) {
+        Order order = orderRepository.findById(id).orElseThrow(() -> new EntityNotFoundException(
+                "Can't find order by id: " + id));
+        order.setStatus(requestDto.getStatus());
+        return orderMapper.toDto(orderRepository.save(order));
+    }
+
+    @Transactional
+    @Override
+    public List<OrderItemResponseDto> getAllByOrderId(Long orderId, Pageable pageable) {
+        return getOrderById(orderId).getOrderItems().stream()
+                .map(orderItemMapper::toDto)
+                .toList();
+    }
+
+    @Transactional
+    @Override
+    public OrderItemResponseDto getOrderItemById(Long orderId, Long itemId) {
+        return getOrderById(orderId).getOrderItems().stream()
+                .filter(c -> Objects.equals(c.getId(), itemId))
+                .map(orderItemMapper::toDto)
+                .findFirst()
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can't find order item by id: " + itemId));
+    }
+
+    private Order getOrderById(Long id) {
+        return orderRepository.findAllByUserId(authentificationService.getUserId()).stream()
+                .filter(o -> Objects.equals(o.getId(), id))
+                .findFirst()
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can't find order by id: " + id));
+    }
+
+    private Set<OrderItem> getOrderItems(ShoppingCart cart, Order order) {
+        return cart.getCartItems().stream()
+                .map(orderItemMapper::mapCartItemToOrderItem)
+                .peek(oi -> oi.setPrice(oi.getPrice().multiply(
```

Comment:
Please don`t use 3 same operations in a row, maybe it is possible to create a function and just call it in one peek?

Reply:
not resolved